protoc-gen-go/testdata/grpc.pb.go

```diff
@@ -0,0 +1,370 @@
+}
+
+// Server API for Test service
+
+type TestServer interface {
+	UnaryCall(context.Context, *SimpleRequest) (*SimpleResponse, error)
+	// This RPC streams from the server only.
+	Downstream(*SimpleRequest, Test_DownstreamServer) error
+	// This RPC streams from the client.
+	Upstream(Test_UpstreamServer) error
+	// This one streams in both directions.
+	Bidi(Test_BidiServer) error
+}
+
+func RegisterTestServer(s *grpc.Server, srv TestServer) {
+	s.RegisterService(&_Test_serviceDesc, srv)
+}
+
+func _Test_UnaryCall_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc.UnaryServerInterceptor) (interface{}, error) {
+	in := new(SimpleRequest)
+	if err := dec(in); err != nil {
+		return nil, err
+	}
+	if interceptor == nil {
+		return srv.(TestServer).UnaryCall(ctx, in)
+	}
+	info := &grpc.UnaryServerInfo{
+		Server:     srv,
+		FullMethod: "/testdata.Test/UnaryCall",
+	}
+	handler := func(ctx context.Context, req interface{}) (interface{}, error) {
+		return srv.(TestServer).UnaryCall(ctx, req.(*SimpleRequest))
+	}
+	return interceptor(ctx, in, info, handler)
+}
+
+func _Test_Downstream_Handler(srv interface{}, stream grpc.ServerStream) error {
+	m := new(SimpleRequest)
+	if err := stream.RecvMsg(m); err != nil {
+		return err
+	}
+	return srv.(TestServer).Downstream(m, &testDownstreamServer{stream})
+}
+
+type Test_DownstreamServer interface {
+	Send(*StreamMsg) error
+	grpc.ServerStream
+}
+
+type testDownstreamServer struct {
+	grpc.ServerStream
+}
+
+func (x *testDownstreamServer) Send(m *StreamMsg) error {
+	return x.ServerStream.SendMsg(m)
+}
+
+func _Test_Upstream_Handler(srv interface{}, stream grpc.ServerStream) error {
+	return srv.(TestServer).Upstream(&testUpstreamServer{stream})
+}
+
+type Test_UpstreamServer interface {
+	SendAndClose(*SimpleResponse) error
+	Recv() (*StreamMsg, error)
+	grpc.ServerStream
+}
+
+type testUpstreamServer struct {
+	grpc.ServerStream
+}
+
+func (x *testUpstreamServer) SendAndClose(m *SimpleResponse) error {
+	return x.ServerStream.SendMsg(m)
+}
+
+func (x *testUpstreamServer) Recv() (*StreamMsg, error) {
+	m := new(StreamMsg)
+	if err := x.ServerStream.RecvMsg(m); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func _Test_Bidi_Handler(srv interface{}, stream grpc.ServerStream) error {
+	return srv.(TestServer).Bidi(&testBidiServer{stream})
+}
+
+type Test_BidiServer interface {
+	Send(*StreamMsg2) error
+	Recv() (*StreamMsg, error)
+	grpc.ServerStream
+}
+
+type testBidiServer struct {
+	grpc.ServerStream
+}
+
+func (x *testBidiServer) Send(m *StreamMsg2) error {
+	return x.ServerStream.SendMsg(m)
+}
+
+func (x *testBidiServer) Recv() (*StreamMsg, error) {
+	m := new(StreamMsg)
+	if err := x.ServerStream.RecvMsg(m); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func GetTestServiceDescriptor() ([]byte, []int) { return fileDescriptor0, []int{0} }
+
+var _Test_serviceDesc = GetTestServiceDesc()
+
+func GetTestServiceDesc() grpc.ServiceDesc {
```

Comment:
grpc.ServiceDesc is intended to be passed around by reference. Current design uses a non-pointer _ServiceName_serviceDesc because of a complicated argument about about behind the scenes allocations.

However, here, we are definitely passing around each individual element directly on the stack every time it is called or used in a function. That’s a lot of copying data around, especially for a particularly large Service.

Reply:
Is this a comment on the current implementation or this PR?

Reply:
Make this return a pointer. I don't know if it particularly matters that this copies data. It's not good, for sure, but ServiceDescs are always referenced for the intents of using the grpc package.

Reply:
Generated grpc code that use the ServiceDesc still use the var up above. It does not get copied. Only callers outside of grpc would get a copy. If this doesn't matter too much, I'd rather not change this, and avoid the awkwardness of doing something like

```
var _Test_serviceDescCopy = GetTestServiceDesc()
var _Test_serviceDesc = &_Test_serviceDescCopy
```

Reply:
@paranoiacblack after reading your comment in the other file again, did you actually want the code to look like this?

```
var _Test_serviceDesc = &grpc.ServiceDesc{
  ...
}

func GetTestServiceDesc() *grpc.ServiceDesc {
	return _Test_serviceDesc
}
```

This would make usages able to mutate the `ServiceDesc` and affect how the service is registered. Not sure that would be desired behavior...

Reply:
What I wanted, explicitly was for the original definition of `_Test_serviceDesc` to remain untouched:

```
var _Test_serviceDesc = grpc.ServiceDesc
```

And for `GetTestServiceDesc` to do the following:

```
func GetTestServiceDesc() *grpc.ServiceDesc {
        return &_Test_serviceDesc
}
```

You are correct that this will allow service descriptors to be modified by the server/clients that wish to register and use them. However, I think that might have been some of the intention behind this change as well. Your use-case involves only introspecting these objects so you can make observations about which methods exists, but another use-case might be adding ad-hoc methods or removing them ad-hoc from a given service before registration (like some sort of middleware interception).

Reply:
That would be asking the user to change global state in order to change how a server is registered. If that's something you want to support, it actually sounds like the support should be built into service registration in `grpc`. cc: @dfawley

Reply:
I think that the way the `grpc` package will support this is by using `GetServiceDesc` and modifying it, probably. Either way, Service Descriptors are meant to be passed around by reference.

Reply:
I would much prefer keeping the _Test_serviceDesc definition alone, but make a copy right before returning the pointer-to.

```
var _Test_serviceDesc = grpc.ServiceDesc{
  ...
}

func GetTestServiceDesc() *grpc.ServiceDesc {
	desc := _Test_serviceDesc
	return &desc
}
```

Yes, this makes a copy, but a) it doesn’t make a copy everytime, b) it’s unavoidable if one wishes to avoid mutability of state, c) should be the uncommon case, and so optimizing it for performance would be the wrong way to do optimization anyways.

However, considering the structure could be quite large, let’s say 64 bytes isn’t unreasonable, then just _using_ it incurs this copy cost every time. So, we copy once as it is leaving our implementation, and then usage continues with pointers.

Reply:
@puellanivis, This doesn't make a deep copy, so if the user changed the slices' contents, that would impact the global state.

If we are doing this (which I'm still not totally convinced is worth doing yet), I think the way it's written is fine, except change the return type of this function to `*grpc.ServiceDesc`.  Then it will be passed around more naturally, but everyone that asks for it gets their own version.  This really should not be mutable.

Reply:
Ugh… oh yeah.

I had looked at an earlier version that was returning a complex literal, and so the the only change would have been adding a "&" instead… I think things have shifted since then.

I’ll just walk away again… :/ (not enough coffee)